File: Visual-Turing-Test-main/visual turing website agransh/scripts/convert-to-png.js
/**
 * Convert GIF and NPY images to PNG for the Visual Turing Test website.
 * Run: npm run convert-to-png
 */
const fs = require('fs');
const path = require('path');

const PUBLIC_DIR = path.join(__dirname, '..', 'public', 'agransh_images');
// Real with patho: 30, 28, 36, 26, 18 from gif slices\Real\Real With Pathology (not mask)
const REAL_WITH_PATHO_NPY_SOURCE = path.join('D:', '1_Agransh', 'gif slices', 'Real', 'Real With Pathology');
const REAL_WITH_PATHO_NPY_FILES = ['30.npy', '28.npy', '36.npy', '26.npy', '18.npy'];

async function convertGifToPng(gifPath, pngPath) {
  try {
    const sharp = require('sharp');
    await sharp(gifPath)
      .png()
      .toFile(pngPath);
    return true;
  } catch (err) {
    console.error(`  Error converting ${path.basename(gifPath)}:`, err.message);
    return false;
  }
}

const NUM_SLICES_PER_SCAN = 8;

/** Normalize slice data to 0-255 and write one PNG. */
async function writeSlicePng(sliceData, height, width, pngPath) {
  const sharp = require('sharp');
  let min = sliceData[0], max = sliceData[0];
  for (let i = 1; i < sliceData.length; i++) {
    if (sliceData[i] < min) min = sliceData[i];
    if (sliceData[i] > max) max = sliceData[i];
  }
  const range = max > min ? max - min : 1;
  const pngBuf = Buffer.alloc(width * height);
  for (let i = 0; i < sliceData.length; i++) {
    pngBuf[i] = Math.round(((sliceData[i] - min) / range) * 255);
  }
  await sharp(pngBuf, { raw: { width, height, channels: 1 } }).png().toFile(pngPath);
}

/**
 * Convert one NPY volume to 8 slice PNGs (same volume, 8 divisions).
 * Output: vol_0.png ... vol_7.png in destFolder.
 * If 3D [N,H,W]: use first 8 slices (or evenly spaced if N>8). If 2D: duplicate slice 8 times.
 */
async function convertNpyTo8Slices(npyPath, destFolder, baseName) {
  try {
    const { load } = require('npyjs');
    const buf = fs.readFileSync(npyPath);
    const arrayBuffer = buf.buffer.slice(buf.byteOffset, buf.byteOffset + buf.byteLength);
    const arr = await load(arrayBuffer);
    let data = arr.data;
    const shape = arr.shape;
    const sharp = require('sharp');

    let numSlices = 1, height, width, sliceSize;
    if (shape.length === 3) {
      numSlices = shape[0];
      height = shape[1];
      width = shape[2];
      sliceSize = height * width;
    } else if (shape.length === 2) {
      height = shape[0];
      width = shape[1];
      sliceSize = height * width;
    } else {
      console.error(`  Unsupported shape ${shape} for ${path.basename(npyPath)}`);
      return false;
    }

    const indices = [];
    if (numSlices >= NUM_SLICES_PER_SCAN) {
      for (let s = 0; s < NUM_SLICES_PER_SCAN; s++) {
        indices.push(Math.floor((s * numSlices) / NUM_SLICES_PER_SCAN));
      }
    } else {
      for (let s = 0; s < NUM_SLICES_PER_SCAN; s++) {
        indices.push(Math.min(s % numSlices, numSlices - 1));
      }
    }

    for (let i = 0; i < NUM_SLICES_PER_SCAN; i++) {
      const sliceIdx = indices[i];
      const start = sliceIdx * sliceSize;
      const sliceData = Array.from(data.slice(start, start + sliceSize));
      const pngPath = path.join(destFolder, `${baseName}_${i}.png`);
      await writeSlicePng(sliceData, height, width, pngPath);
    }
    return true;
  } catch (err) {
    console.error(`  Error converting ${path.basename(npyPath)}:`, err.message);
    return false;
  }
}

async function main() {
  console.log('Converting images to PNG...\n');

  // 1. Real with patho: each NPY -> 8 slice PNGs (same volume, 8 divisions) e.g. 30_0.png ... 30_7.png
  const realWithPathoDest = path.join(PUBLIC_DIR, 'Real_with_patho');
  if (!fs.existsSync(realWithPathoDest)) fs.mkdirSync(realWithPathoDest, { recursive: true });
  if (fs.existsSync(REAL_WITH_PATHO_NPY_SOURCE)) {
    console.log('Real_with_patho (8 slices per NPY volume):');
    for (const npyName of REAL_WITH_PATHO_NPY_FILES) {
      const npyPath = path.join(REAL_WITH_PATHO_NPY_SOURCE, npyName);
      if (fs.existsSync(npyPath)) {
        const baseName = path.basename(npyName, '.npy');
        if (await convertNpyTo8Slices(npyPath, realWithPathoDest, baseName)) {
          console.log(`  ${npyName} -> ${baseName}_0.png ... ${baseName}_7.png (8 slices same volume)`);
        }
      } else {
        console.log(`  Skipping (not found): ${npyName}`);
      }
    }
    console.log('');
  } else {
    console.log(`Skipping Real_with_patho NPY (folder not found): ${REAL_WITH_PATHO_NPY_SOURCE}\n`);
  }

  // 2. GIF to PNG in existing public folders
  const folders = ['Real_wo_patho', 'Generated', 'Real_with_patho'];
  for (const folder of folders) {
    const folderPath = path.join(PUBLIC_DIR, folder);
    if (!fs.existsSync(folderPath)) continue;
    const files = fs.readdirSync(folderPath).filter((f) => f.endsWith('.gif'));
    if (files.length === 0) continue;
    console.log(`${folder} (GIF -> PNG):`);
    for (const file of files) {
      const gifPath = path.join(folderPath, file);
      const pngName = path.basename(file, '.gif') + '.png';
      const pngPath = path.join(folderPath, pngName);
      if (await convertGifToPng(gifPath, pngPath)) console.log(`  ${file} -> ${pngName}`);
    }
    console.log('');
  }

  console.log('Done! PNG images saved to public/agransh_images/');
}

main().catch(console.error);

File: Visual-Turing-Test-main/visual turing website agransh/scripts/convert_to_png.py
"""
Convert GIF and NPY images to PNG for the Visual Turing Test website.
Run from project root: python scripts/convert_to_png.py
"""
import os
import sys
from pathlib import Path

# Source and destination paths
SCRIPT_DIR = Path(__file__).parent
PROJECT_DIR = SCRIPT_DIR.parent
PUBLIC_DIR = PROJECT_DIR / "public" / "agransh_images"
SOURCE_BASE = Path(r"D:\1_Agransh\gif slices")

# Folder mappings: source subfolder -> destination folder name
FOLDERS = [
    (SOURCE_BASE / "Real" / "Real wo patho", "Real_wo_patho"),
    (SOURCE_BASE / "Generated", "Generated"),
]

# Real with patho - from gif files to send to arjun (legacy)
REAL_WITH_PATHO_SOURCE = Path(r"D:\1_Agransh\gif files to send to arjun")
# Real with patho NPY files (30, 28, 36, 26, 18) from gif slices\Real\Real With Pathology
REAL_WITH_PATHO_NPY_SOURCE = Path(r"D:\1_Agransh\gif slices\Real\Real With Pathology")
REAL_WITH_PATHO_NPY_FILES = ["30.npy", "28.npy", "36.npy", "26.npy", "18.npy"]


def convert_gif_to_png(src_path, dest_path):
    """Convert GIF to PNG using PIL."""
    try:
        from PIL import Image
        img = Image.open(src_path)
        if img.mode in ("RGBA", "P"):
            img = img.convert("RGB")
        img.save(dest_path, "PNG")
        return True
    except Exception as e:
        print(f"  Error converting {src_path.name}: {e}")
        return False


def convert_npy_to_png(src_path, dest_path):
    """Convert NPY array to PNG using numpy and PIL."""
    try:
        import numpy as np
        from PIL import Image
        arr = np.load(src_path)
        if arr.ndim == 3:
            arr = arr.squeeze()
        if arr.ndim == 2:
            pass
        elif arr.ndim == 3 and arr.shape[-1] in (1, 3):
            arr = arr.squeeze()
        else:
            arr = arr[0] if arr.ndim > 2 else arr
        # Normalize to 0-255
        arr_min, arr_max = arr.min(), arr.max()
        if arr_max > arr_min:
            arr = ((arr - arr_min) / (arr_max - arr_min) * 255).astype(np.uint8)
        else:
            arr = np.zeros_like(arr, dtype=np.uint8)
        img = Image.fromarray(arr)
        img.save(dest_path, "PNG")
        return True
    except Exception as e:
        print(f"  Error converting {src_path.name}: {e}")
        return False


def main():
    print("Converting images to PNG...")
    
    for src_folder, dest_name in FOLDERS:
        if not src_folder.exists():
            print(f"Skipping (not found): {src_folder}")
            continue
        dest_folder = PUBLIC_DIR / dest_name
        dest_folder.mkdir(parents=True, exist_ok=True)
        print(f"\n{dest_name}:")
        
        for f in sorted(src_folder.iterdir()):
            if f.suffix.lower() == ".gif":
                out_name = f.stem + ".png"
                out_path = dest_folder / out_name
                if convert_gif_to_png(f, out_path):
                    print(f"  {f.name} -> {out_name}")
            elif f.suffix.lower() == ".npy":
                out_name = f.stem + ".png"
                out_path = dest_folder / out_name
                if convert_npy_to_png(f, out_path):
                    print(f"  {f.name} -> {out_name}")
    
    # Real with patho - NPY files 30, 28, 36, 26, 18 from mask folder
    dest_folder = PUBLIC_DIR / "Real_with_patho"
    dest_folder.mkdir(parents=True, exist_ok=True)
    print(f"\nReal_with_patho (from NPY):")
    if REAL_WITH_PATHO_NPY_SOURCE.exists():
        for npy_name in REAL_WITH_PATHO_NPY_FILES:
            f = REAL_WITH_PATHO_NPY_SOURCE / npy_name
            if f.exists():
                out_name = f.stem + ".png"
                out_path = dest_folder / out_name
                if convert_npy_to_png(f, out_path):
                    print(f"  {f.name} -> {out_name}")
            else:
                print(f"  Skipping (not found): {f}")
    else:
        print(f"  Skipping (folder not found): {REAL_WITH_PATHO_NPY_SOURCE}")

    # Real with patho - legacy sample gifs (optional)
    if REAL_WITH_PATHO_SOURCE.exists():
        for f in sorted(REAL_WITH_PATHO_SOURCE.glob("*_sample.gif")):
            out_name = f.stem + ".png"
            out_path = dest_folder / out_name
            if convert_gif_to_png(f, out_path):
                print(f"  {f.name} -> {out_name}")
    
    print("\nDone! PNG images saved to public/agransh_images/")


if __name__ == "__main__":
    main()
